File: api/src/main/java/org/pmiops/workbench/actionaudit/Agent.java
package org.pmiops.workbench.actionaudit;

import jakarta.annotation.Nullable;
import org.pmiops.workbench.db.model.DbUser;

public record Agent(AgentType agentType, @Nullable Long idMaybe, @Nullable String emailMaybe) {
  public static Agent asUser(DbUser u) {
    return new Agent(AgentType.USER, u.getUserId(), u.getUsername());
  }

  public static Agent asAdmin(DbUser u) {
    return new Agent(AgentType.ADMINISTRATOR, u.getUserId(), u.getUsername());
  }

  public static Agent asSystem() {
    return new Agent(AgentType.SYSTEM, null, null);
  }
}

File: api/src/main/java/org/pmiops/workbench/actionaudit/Agent.kt
package org.pmiops.workbench.actionaudit

import org.pmiops.workbench.db.model.DbUser

data class Agent constructor(
    val agentType: AgentType,
    val idMaybe: Long? = null,
    val emailMaybe: String? = null
) {

    companion object {
        @JvmStatic
        fun asUser(u: DbUser): Agent {
            return Agent(
                    agentType = AgentType.USER,
                    idMaybe = u.userId,
                    emailMaybe = u.username)
        }

        @JvmStatic
        fun asAdmin(u: DbUser): Agent {
            return Agent(
                    agentType = AgentType.ADMINISTRATOR,
                    idMaybe = u.userId,
                    emailMaybe = u.username)
        }

        @JvmStatic
        fun asSystem(): Agent {
            return Agent(agentType = AgentType.SYSTEM)
        }
    }
}